### PR TITLE
RM-7870: The return values of Path.GetDirectoryName must be checked for null in Remotion.Security.Metadata.LocalizationFileNameStrategy

### PR DESCRIPTION
**Pull request by raffaelfoidl**
https://re-motion.atlassian.net/browse/RM-7870

### CODE REVIEW

File: Remotion/Security/Core/Metadata/LocalizationFileNameStrategy.cs [hunk 1]
```diff
@@ -27,8 +28,9 @@ public class LocalizationFileNameStrategy
     public string GetLocalizationFileName (string metadataFilename, CultureInfo culture)
     {
       string baseFilename = Path.GetFileNameWithoutExtension (metadataFilename);
-      // TODO RM-7870: Add notnull assertion
-      string basePath = Path.GetDirectoryName (metadataFilename)!;
+      string? basePath = Path.GetDirectoryName (metadataFilename);
+      Assertion.IsNotNull (basePath);
```
Comment (iam-mholle) [inline Remotion/Security/Core/Metadata/LocalizationFileNameStrategy.cs:32]: Please add a message, maybe something of the sort `nameof(metadataFilename) + " must contain a directory name."`? @MichaelKetting

**Review by iam-mholle — CHANGES_REQUESTED** [review]

File: Remotion/Security/Core/Metadata/LocalizationFileNameStrategy.cs (hunk 1 above)
Comment (raffaelfoidl) [reply Remotion/Security/Core/Metadata/LocalizationFileNameStrategy.cs]: See 79d79dad6011aaf8d60c7e29251a69c5e3a363b3. Went with "non-root directory" due to respective docs.

Comment (MichaelKetting) [reply Remotion/Security/Core/Metadata/LocalizationFileNameStrategy.cs]: Thanks to the acutal code and discussion and need to inject the parameter value, I realized that a) the method is public, b) this is a parameter, and c) the paramter's value is actually invalid if this error occurs. So, we should make this into a straight forward argument exception.

File: Remotion/Security/Core/Metadata/LocalizationFileNameStrategy.cs [hunk 2]
```diff
@@ -27,8 +28,9 @@ public class LocalizationFileNameStrategy
     public string GetLocalizationFileName (string metadataFilename, CultureInfo culture)
     {
       string baseFilename = Path.GetFileNameWithoutExtension (metadataFilename);
-      // TODO RM-7870: Add notnull assertion
-      string basePath = Path.GetDirectoryName (metadataFilename)!;
+      string? basePath = Path.GetDirectoryName (metadataFilename);
+      Assertion.IsNotNull (basePath, nameof (metadataFilename) + " must denote a path that is not a root directory.");
```
Comment (MichaelKetting) [inline Remotion/Security/Core/Metadata/LocalizationFileNameStrategy.cs:32]: See discussion.

**Review by MichaelKetting — CHANGES_REQUESTED** [review]

File: Remotion/Security/Core/Metadata/LocalizationFileNameStrategy.cs [hunk 3]
```diff
@@ -30,8 +30,13 @@ public string GetLocalizationFileName (string metadataFilename, CultureInfo cult
...
 
       string baseFilename = Path.GetFileNameWithoutExtension (metadataFilename);
-      // TODO RM-7870: Add notnull assertion
-      string basePath = Path.GetDirectoryName (metadataFilename)!;
+
+      string? basePath = Path.GetDirectoryName (metadataFilename);
+      if (basePath == null)
+      {
+        throw new ArgumentException ("The metadata filename must denote a path that is not a root directory.", "metadataFilename");
+      }
```
Comment (MichaelKetting) [inline Remotion/Security/Core/Metadata/LocalizationFileNameStrategy.cs:38]:
```suggestion
        throw new ArgumentException ("The metadata filename must denote a path that is not a root directory.", "metadataFilename");
```

For single-line IFs, we can (and usually should) sckip the branches to improve readdability/compactness of the code.

File: Remotion/Security/Core/Metadata/LocalizationFileNameStrategy.cs [hunk 4]
```diff
@@ -45,8 +50,13 @@ public string[] GetLocalizationFileNames (string metadataFilename)
...
 
       string baseFileName = Path.GetFileNameWithoutExtension (metadataFilename);
-      // TODO RM-7870: Add notnull assertion
-      string basePath = Path.GetDirectoryName (metadataFilename)!;
+
+      string? basePath = Path.GetDirectoryName (metadataFilename);
+      if (basePath == null)
+      {
+        throw new ArgumentException ("The metadata filename must denote a path that is not a root directory.", "metadataFilename");
+      }
```
Comment (MichaelKetting) [inline Remotion/Security/Core/Metadata/LocalizationFileNameStrategy.cs:58]:
```suggestion
        throw new ArgumentException ("The metadata filename must denote a path that is not a root directory.", "metadataFilename");
```

**Review by MichaelKetting — CHANGES_REQUESTED** [review]

**Review by MichaelKetting — APPROVED** [review]

**Review by iam-mholle — APPROVED** [review]

**Pull request merged**